### PR TITLE
Do not allocate memory for DoFs on artificial cells

### PR DESCRIPTION
**Pull request by peterrum**
This reduces the memory consumption of `DoFHandler` for a non-refined `p:d:T` from `#cells * #n_dofs_cells` to `#cells + #non_artificial_cells * #n_dofs_cells`. The `#cells` (pointers) is still not optimal; but is caused by the fact that all coarse cells are stored and the way cell information are accessed (cell level and index).

### CODE REVIEW

File: include/deal.II/dofs/dof_accessor.templates.h [hunk 1]
```diff
@@ -811,9 +791,10 @@ namespace internal
       n_dof_indices(
         const dealii::DoFAccessor<structdim, dim, spacedim, level_dof_access>
           &                accessor,
-        const unsigned int fe_index_)
+        const unsigned int fe_index_,
+        const bool         level_dof_access_)
```
Comment (kronbichler) [inline include/deal.II/dofs/dof_accessor.templates.h:795]: Do get it right that you change this to be able to use this new parameter independently of the template parameter of `DoFAccessor`? Can you maybe add a more descriptive name, say `count_level_dofs` or something like that? (And the same in the function using this below.)

**Review by kronbichler — APPROVED** [review]
I like this a lot, thanks for the work. 

The old approach to allocate the memory was just insane, but we were forced to in the original data structures. Thanks to the big updates that brought the `hp::DoFHandler` into `DoFHandler` (among other things), we have gained the flexibility to finally get rid of this nuisance.

File: include/deal.II/dofs/dof_accessor.templates.h (hunk 1 above)
Comment (peterrum) [reply include/deal.II/dofs/dof_accessor.templates.h]: Done :+1:

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/dofs/dof_accessor.templates.h [hunk 2]
```diff
@@ -811,9 +791,13 @@ namespace internal
       n_dof_indices(
         const dealii::DoFAccessor<structdim, dim, spacedim, level_dof_access>
           &                accessor,
-        const unsigned int fe_index_)
+        const unsigned int fe_index_,
+        const bool         count_level_dofs)
       {
-        if (level_dof_access)
+        // note: we cannot rely on the the template parameter level_dof_access
```
Comment (tjhei) [inline include/deal.II/dofs/dof_accessor.templates.h:797]: Duplicated the

Comment (peterrum) [reply include/deal.II/dofs/dof_accessor.templates.h:797]: Done!

**Pull request merged**